cpp/src/search/contains.cu

```diff
@@ -56,15 +56,21 @@ struct contains_scalar_dispatch {
     if (haystack.has_nulls()) {
       auto const begin = d_haystack->pair_begin<DType, true>();
       auto const end   = d_haystack->pair_end<DType, true>();
-      auto const val   = thrust::make_pair(s->value(stream), true);
 
-      return thrust::find(rmm::exec_policy(stream), begin, end, val) != end;
+      return thrust::count_if(rmm::exec_policy(stream),
+                              begin,
+                              end,
+                              [needle_pair = thrust::make_pair(s->value(stream), true)] __device__(
+                                auto const val_pair) { return val_pair == needle_pair; });
     } else {
       auto const begin = d_haystack->begin<DType>();
       auto const end   = d_haystack->end<DType>();
-      auto const val   = s->value(stream);
 
-      return thrust::find(rmm::exec_policy(stream), begin, end, val) != end;
+      return thrust::count_if(
+        rmm::exec_policy(stream),
+        begin,
+        end,
+        [needle = s->value(stream)] __device__(auto const val) { return val == needle; });
```

Comment:
Should we pass a reference here (and other equivalent cases)?
```suggestion
        [needle = s->value(stream)] __device__(auto const& val) { return val == needle; });
```

Reply:
As discussed recently, we should not pass reference in device code: https://github.com/rapidsai/cudf/pull/11019#discussion_r899467478

Reply:
The `needle = s->value(stream)` does seem wasteful here since it is copying the value of `s` from device-to-host only to pass it by value to the lambda. Perhaps this copy can be avoided?
```
auto d_needle = get_scalar_device_view(*s);
 [d_needle] __device__(auto const& val) { return val == needle.value(); });
```
I may have the syntax wrong.

Reply:
It's not obvious to me which approach would be faster. The cost of the device-host copy is one factor. Another factor that could benefit the current approach is that the `needle` parameter, if known by the host and passed into the kernel, might be treated as a kernel-wide constant that is stored in a more "local" place in the memory hierarchy. I can't guess what the compiler will think here but I would benchmark both approaches. Very curious to see what comes out.

(Also I think we still want to pass `val` by value instead of by reference.)

Reply:
Here is the results after using @davidwendt 's suggestion:
```
Benchmark                                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------------
Contains/SearchScalar_AllValid/32768/manual_time                    -0.2442         -0.1618             0             0             0             0
Contains/SearchScalar_AllValid/262144/manual_time                   -0.1621         -0.1045             0             0             0             0
Contains/SearchScalar_AllValid/2097152/manual_time                  -0.1055         -0.0890             0             0             0             0
Contains/SearchScalar_AllValid/16777216/manual_time                 -0.0263         -0.0288             0             0             0             0
Contains/SearchScalar_AllValid/134217728/manual_time                -0.0040         -0.0043             1             1             1             1
Contains/SearchScalar_AllValid/268435456/manual_time                -0.0098         -0.0522             2             2             2             2
Contains/SearchScalar_Nulls/32768/manual_time                       -0.2641         -0.1822             0             0             0             0
Contains/SearchScalar_Nulls/262144/manual_time                      -0.1597         -0.0973             0             0             0             0
Contains/SearchScalar_Nulls/2097152/manual_time                     -0.0902         -0.0736             0             0             0             0
Contains/SearchScalar_Nulls/16777216/manual_time                    -0.0282         -0.0235             0             0             0             0
Contains/SearchScalar_Nulls/134217728/manual_time                   -0.0073         +0.0147             1             1             1             1
Contains/SearchScalar_Nulls/268435456/manual_time                   -0.0020         +0.0030             2             2             2             2
```

Definitely avoiding a round trip in memory device => host => device is better.